src/KeyTable.cpp

```diff
@@ -0,0 +1,222 @@
+        index = toCharKeysIndex ( --i );
+        ss << &line [21];  // 1st keysym starts at index 21 (skip "\tshift\tcode XXX = " or
+                           // "\taltgr\tcode XXX = ")
+        ss >> std::hex >> utf8code;
+        if ( line [21] == '+' && ( utf8code & 0xB00 ) )
+            utf8code ^= 0xB00;  // see line 0XB00CLUELESS
+
+        if ( line [1] == 's' )  // if line starts with "shift"
+            m_ShiftKeys [index] = static_cast<wchar_t> ( utf8code );
+        if ( line [1] == 'a' )  // if line starts with "altgr"
+            m_AltgrKeys [index] = static_cast<wchar_t> ( utf8code );
+    }  // while (getline(dump, line))
+}
+
+bool hemiola::KeyTable::isCharKey ( unsigned int code ) const
+{
+    if ( code > m_CharOrFunc.size() ) {
+        return false;
+    }
+    return ( m_CharOrFunc [code] == 'c' );
+}
+
+bool hemiola::KeyTable::isFuncKey ( unsigned int code ) const
+{
+    if ( code > m_CharOrFunc.size() ) {
+        return false;
+    }
+    return ( m_CharOrFunc [code] == 'f' );
+}
+
+bool hemiola::KeyTable::isUsedKey ( unsigned int code ) const
+{
+    if ( code > m_CharOrFunc.size() ) {
+        return false;
+    }
+    return ( m_CharOrFunc [code] != '_' );
+}
+
+int hemiola::KeyTable::toCharKeysIndex ( unsigned int code ) const
+{
+    int index = -1;                            // not character code
+    if ( code >= KEY_1 && code <= KEY_EQUAL )  // codes 2-13: US keyboard: 1, 2, ..., 0, -, =
+        index = code - 2;
+    if ( code >= KEY_Q && code <= KEY_RIGHTBRACE )  // codes 16-27: q, w, ..., [, ]
+        index = code - 4;
+    if ( code >= KEY_A && code <= KEY_GRAVE )  // codes 30-41: a, s, ..., ', `
+        index = code - 6;
+    if ( code >= KEY_BACKSLASH && code <= KEY_SLASH )  // codes 43-53: \, z, ..., ., /
+        index = code - 7;
+
+    if ( code == KEY_102ND )
+        index = 47;  // key right to the left of 'Z' on US layout
+
+    return index;
+}
+
+inline int hemiola::KeyTable::toFuncKeysIndex ( unsigned int code ) const
+{
+    int index = -1;         // not character code
+    if ( code == KEY_ESC )  // 1
+        index = 0;
+    if ( code >= KEY_BACKSPACE && code <= KEY_TAB )  // 14-15
+        index = code - 13;
+    if ( code >= KEY_ENTER && code <= KEY_LEFTCTRL )  // 28-29
+        index = code - 25;
+    if ( code == KEY_LEFTSHIFT )
+        index = code - 37;                              // 42
+    if ( code >= KEY_RIGHTSHIFT && code <= KEY_KPDOT )  // 54-83
+        index = code - 48;
+    if ( code >= KEY_F11 && code <= KEY_F12 )  // 87-88
+        index = code - 51;
+    if ( code >= KEY_KPENTER && code <= KEY_DELETE )  // 96-111
+        index = code - 58;
+    if ( code == KEY_PAUSE )  // 119
+        index = code - 65;
+    if ( code >= KEY_LEFTMETA && code <= KEY_COMPOSE )  // 125-127
+        index = code - 70;
+    return index;
+}
+
+// std::optional<wchar_t> hemiola::KeyTable::handleScanCode ( unsigned int code, const KeyState&
```

Comment:
This is only commented out util I begin the key capturing, but I didn't want to think about how to rewrite it just yet.